### ISSUE TITLE
http header recommendations from Nessus scan

### ISSUE THREAD

**Issue by smarquard** [opencast/opencast#1162]
The following emerged from a Nessus scan of an Opencast installation:

Content Security Policy (CSP) Missing
Website Does Not Implement HSTS Best Practices
Website does not implement X-Content-Type-Options Best Practices
Website does not implement X-Frame-Options Best Practices
Website does not implement X-XSS-Protection Best Practices

References:

https://en.wikipedia.org/wiki/Content_Security_Policy
https://en.wikipedia.org/wiki/HTTP_Strict_Transport_Security
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Content-Type-Options
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-XSS-Protection

These need evaluation to see to what extent they're applicable or valuable for an Opencast installation, and if so which should be set by default or noted in documentation.

**Comment by lkiesow** [comment]
Most of this is just configuration and most of this is not true for e.g. our test servers even though I specifically allowed some unsafe options to make development easier:

```
% curl -I https://develop.opencast.org
HTTP/2 302 
server: nginx
date: Wed, 23 Oct 2019 12:22:06 GMT
set-cookie: JSESSIONID=node01krdtukalpep1197iqplu3285q41494.node0;Path=/; HTTPOnly; Secure
expires: Thu, 01 Jan 1970 00:00:00 GMT
location: https://develop.opencast.org/admin-ng/login.html;jsessionid=node01krdtukalpep1197iqplu3285q41494.node0
strict-transport-security: max-age=31536000; includeSubdomains;
x-frame-options: SAMEORIGIN
x-content-type-options: nosniff
x-xss-protection: 1; mode=block
content-security-policy: default-src 'unsafe-inline' 'self'; script-src 'unsafe-inline' 'unsafe-eval' 'self'; font-src 'self' data:; frame-ancestors 'none';
referrer-policy: strict-origin-when-cross-origin
access-control-allow-methods: GET, POST
access-control-allow-credentials: true
access-control-allow-headers: Origin,Content-Type,Accept,Authorization
```

The configuration for this is public in the infrastructure repository as well as ELAN e.V.'s base Ansible scripts.

**Comment by lkiesow** [comment]
More than that, [even our documentations](https://docs.opencast.org/r/8.x/admin/#configuration/https/#general-recommendations) makes recommendations about this since most of this is mentioned e.g. by Mozilla's Observatory

**Comment by smarquard** [comment]
https://en.wikipedia.org/wiki/Secure_by_design
https://en.wikipedia.org/wiki/Secure_by_default

Firstly, like #1161, we should set these in the application where possible rather than relying on installers to set a proxy configuration,

Secondly, it is not relevant that you have added a set of custom options to develop.opencast.org because those custom options are not referenced anywhere in the Opencast documentation (e.g. https://docs.opencast.org/develop/admin/#configuration/https/nginx/) and adopters are unlikely to discover them in the infrastructure repo, nor is it reasonable to expect them to, nor does a reference to Observatory or SSLabs excuse us from doing the right thing by providing secure defaults.

**Comment by lkiesow** [comment]
I like to argue that it makes way more sense for us to urge people to make use of the Mozilla Observatory which explains for non-experts, in detail, step-by-step how to configure your server properly for your use-case than to make recommendations trying to keep up-to-date with the current state of the art security ourselves. That's what our documentation **is** doing right now.

That said if you want to add some of these headers directly to Opencast, feel free to do so. As long as they don't break anything, I'm not opposed to adding those.

I am convinced, however, that you cannot maintain a decent set-up using just Opencast in our current state and we will still need to rely on the Apache httpd, Nginx or other proxies.

**Comment by smarquard** [comment]
I agree it's a moving target and I will take a look at whether the headers can be set internally or not. I do think we need to align our Opencast installation documentation with the configuration for develop.opencast.org (so the nginx config in both is substantially the same, for example), firstly so that adopters have something to start from that is reasonable (and will avoid alerts from nessus and other automated scanning systems), and secondly so that adopters know which options are safe to use (i.e. won't break anything in a standard Opencast setup).

**Comment by lkiesow** [comment]
When writing the guide I deliberately avoided that and instead linked Observatory since they do exactly that and a lot of the options (not all, that's true) highly depend on your system and your integrations. We should not add any settings just to silence a security/vulnerability scanner for the sake of them being quiet. That would defeat the purpose of these scanners ;-)

I guess I can look through and prepare some kind of guide in the next weeks.

**Comment by smarquard** [comment]
This is what happens:
* IT Security team runs a scan on local services.
* Nessus complains about Opencast.
* IT Security forwards the complaints about Opencast to the Opencast admins for action.
* Opencast admins have 2 questions, which they would like answered with a minimum of work (which excludes consulting 3rd-party sources and/or becoming a web security expert):
1. How do I fix these problems? (add headers, etc.)
2. Will fixing them break some aspect of Opencast?

Adding the set of known-not-to-break-Opencast headers, the same ones running on develop.opencast.org, to the default install documentation (for nginx proxy config) address this requirement.

We're only defeating the purpose of a security scanner if we're masking the issue in some way without actually addressing the issue. But we're not doing that, we're actually adding the things that the security scanner is complaining about. If there are some headers that should not be added for every Opencast installation because they may break some integration or aspect of operation, then we document that and the admins can take a calculated risk.

**Comment by smarquard** [comment]
For reference, demo nginx config is here:

https://github.com/opencast/opencast-project-infrastructure/blob/master/ansible-allinone-demo-vm/roles/nginx/templates/nginx.conf

**Comment by karendolan** [comment]
FYI "X-Frame-Options" is deprecated and should be replaced with “Content-Security-Policy”
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options
Example: 
``Content-Security-Policy "frame-ancestors ‘self’ 'https://*.my-trusted-external-site.edu' 'http://localhost:*' 'https://localhost:*';”;``
